atlasdb-config/src/test/java/com/palantir/atlasdb/http/FailoverFeignTargetTest.java

```diff
@@ -0,0 +1,136 @@
+    private static final int ITERATIONS = 100;
+    private static final int CLUSTER_SIZE = 3;
+
+    private static final String SERVER_1 = "server1";
+    private static final String SERVER_2 = "server2";
+    private static final String SERVER_3 = "server3";
+    private static final List<String> SERVERS = ImmutableList.of(SERVER_1, SERVER_2, SERVER_3);
+
+    private static final RetryableException EXCEPTION_WITH_RETRY_AFTER = mock(RetryableException.class);
+    private static final RetryableException EXCEPTION_WITHOUT_RETRY_AFTER = mock(RetryableException.class);
+    private static final RetryableException BLOCKING_TIMEOUT_EXCEPTION = mock(RetryableException.class);
+
+    private final FailoverFeignTarget<Object> target = new FailoverFeignTarget<>(
+            SERVERS, 1, Object.class);
+
+    static {
+        when(EXCEPTION_WITH_RETRY_AFTER.retryAfter()).thenReturn(Date.valueOf(LocalDate.MAX));
+        when(EXCEPTION_WITHOUT_RETRY_AFTER.retryAfter()).thenReturn(null);
+        when(BLOCKING_TIMEOUT_EXCEPTION.getCause()).thenReturn(new BlockingTimeoutException(new Exception()));
+    }
+
+    @Test
+    public void failsOverOnExceptionWithRetryAfter() {
+        String initialUrl = target.url();
+        target.continueOrPropagate(EXCEPTION_WITH_RETRY_AFTER);
+        assertThat(target.url()).isNotEqualTo(initialUrl);
+    }
+
+    @Test
+    public void failsOverMultipleTimesOnExceptionsWithRetryAfter() {
+        String previousUrl;
+        for (int i = 0; i < FAILOVERS; i++) {
+            previousUrl = target.url();
+            target.continueOrPropagate(EXCEPTION_WITH_RETRY_AFTER);
+            assertThat(target.url()).isNotEqualTo(previousUrl);
+        }
+    }
+
+    @Test
+    public void doesNotImmediatelyFailOverOnExceptionWithoutRetryAfter() {
+        String initialUrl = target.url();
+        target.continueOrPropagate(EXCEPTION_WITHOUT_RETRY_AFTER);
+        assertThat(target.url()).isEqualTo(initialUrl);
+    }
+
+    @Test
+    public void rethrowsExceptionWithoutRetryAfterWhenLimitExceeded() {
+        assertThatThrownBy(() -> {
+            for (int i = 0; i < FAILOVERS; i++) {
+                target.url();
+                target.continueOrPropagate(EXCEPTION_WITHOUT_RETRY_AFTER);
+            }
+        }).isEqualTo(EXCEPTION_WITHOUT_RETRY_AFTER);
+    }
+
+    @Test
+    public void doesNotFailOverOnBlockingTimeoutException() {
+        String initialUrl = target.url();
+        target.continueOrPropagate(BLOCKING_TIMEOUT_EXCEPTION);
+        assertThat(target.url()).isEqualTo(initialUrl);
+    }
+
+    @Test
+    public void doesNotFailOverOnMultipleBlockingTimeoutExceptions() {
+        String initialUrl = target.url();
+        for (int i = 0; i < FAILOVERS; i++) {
+            target.continueOrPropagate(BLOCKING_TIMEOUT_EXCEPTION);
+            assertThat(target.url()).isEqualTo(initialUrl);
+        }
+    }
+
+    @Test
+    public void failsOverMultipleTimesWithFailingLeader() {
```

Comment:
The number of times to retry in total isn't actually CLUSTER_SIZE * failuresBeforeSwitching though, it's hard-coded at 14 currently. I'll rewrite this to avoid using modulus at all (switch on the value of the url instead).